# Parte 2 - Programación con Scala - Ejercicios 01 a 05

## Ejercicio 1

### Versión A, con el tipo de variable declarado explícitamente
Datos ficticios:
- Nombre del jugador: Marco
- Número de partidas jugadas: 10
- Puntuación actual: 73
- Porcentaje de victorias: 84 (lo pondré como float de 0 a 1)
- Indicación de si el jugador continúa activo: true

In [12]:
val nombre: String = "Marco"
val partidas: Int = 10
val puntuacion: Int = 73
val porcentajeVictorias: Float = 0.84f
val jugadorActivo: Boolean = true

println(nombre)
println(partidas)
println(puntuacion)
println(porcentajeVictorias)
println(jugadorActivo)

Marco
10
73
0.84
true


nombre: String = "Marco"
partidas: Int = 10
puntuacion: Int = 73
porcentajeVictorias: Float = 0.84F
jugadorActivo: Boolean = true

### Versión B, con el tipo de variable sin declarar explícitamente (Scala los infiere)
Datos ficticios:
- Nombre del jugador: Marco
- Número de partidas jugadas: 10
- Puntuación actual: 73
- Porcentaje de victorias: 84 (lo pondré como float de 0 a 1)
- Indicación de si el jugador continúa activo: true

In [13]:
val nombre = "Marco"
val partidas = 10
val puntuacion = 73
val porcentajeVictorias = 0.84f
val jugadorActivo = true

println(nombre)
println(partidas)
println(puntuacion)
println(porcentajeVictorias)
println(jugadorActivo)

Marco
10
73
0.84
true


nombre: String = "Marco"
partidas: Int = 10
puntuacion: Int = 73
porcentajeVictorias: Float = 0.84F
jugadorActivo: Boolean = true

### Explicación
- En la versión A, Scala no ha inferido ningún tipo, puesto que yo mismo he declarado el tipo de cada variable, mientras que en la versión B, Scala ha inferido los siguientes tipos:
    - nombre: String
    - partidas: Int
    - puntuacion: Int
    - porcentajeVictorias: Float
    - jugadorActivo: Boolean
- La única diferencia que observo es que en la versión A, puedo elegir el tipo de dato que yo quiera para cada variable (Ejemplo: porcentajeVictorias podría ser un Int de 0 a 100, en lugar de ser un float de 0 a 1), mientras que en la versión B, el propio Scala asigna el tipo de variable que considera oportuno.
- Las variables que deberían ser val o var:
    - nombre: val (a no ser que se contemple la posibilidad de que el nombre esté escrito mal y se quiera modificar)
    - partidas: var (la variable aumenta cada partida que el jugador juega)
    - puntuacion: var (la puntuación varía en función de la habilidad del jugador)
    - porcentajeVictorias: var (el porcentaje varía cada vez que una partida finaliza)
    - jugadorActivo: var (el estado del jugador cambia si este está activo o no)

## Ejercicio 2

Se tienen 2 variables las cuales se pide que se modifiquen o intenten modificar:
- jugador (inmutable) --> Dará error si se trata de modificar
- puntuación (mutable) --> Se modificará sin problema

In [14]:
val jugador = "Marta"
var puntuacion = 10

puntuacion += 5
puntuacion += 3
println(jugador)
println(puntuacion)
//jugador = "Marco" //Comento la línea para que no de error. Si se descomenta dará error

Marta
18


jugador: String = "Marta"
puntuacion: Int = 18

In [14]:
// DEMOSTRACIÓN DEL ERROR QUE DEBERÍA SALIR
jugador = "Marco"

cmd15.sc:2: reassignment to val
jugador = "Marco"
        ^
Compilation Failed

### Explicación
- La puntuación es modificable porque es una variable "var", que es mutable (permite modificaciones)
- El nombre del jugador no se puede modificar porque es una variable "val", que es inmutable (no permite modificaciones)
- Scala genera el siguiente mensaje si se trata de modificar un val:
    ```
    cmd4.sc:2: reassignment to val
    jugador = "Marco"
            ^
    Compilation Failed
    ```

## Ejercicio 3

Se pide declarar un valor decimal con Double, y el mismo valor con Float.
Dicho valor será: 3.14159265358979323846264338327

Después, se pide crear una variable Int, otra Boolean, y otra String.

In [15]:
val variableDouble: Double = 3.14159265358979323846264338327
val variableFloat: Float = 3.14159265358979323846264338327f
val variableInt: Int = 4
val variableBoolean: Boolean = false
val variableString: String = "Hola"

println(variableDouble)
println(variableFloat)
println(variableInt)
println(variableBoolean)
println(variableString)

3.141592653589793
3.1415927
4
false
Hola


variableDouble: Double = 3.141592653589793
variableFloat: Float = 3.1415927F
variableInt: Int = 4
variableBoolean: Boolean = false
variableString: String = "Hola"

### Explicación
- La diferencia entre Double y Float es que el Float se trunca en el séptimo decimal, permitiendo únicamente representar 3.1415927. Por otra parte, el Double se trunca en el decimoquinto decimal, permitiendo únicamente representar 3.141592653589793.
      - El número que se pretende representar es 3.14159265358979323846264338327
- Para el número de estudiantes de una clase, utilizaría una variable de tipo Int, puesto que siempre será un número entero.
- Para representar si un estudiante ha aprobado o no, utilizaría una variable de tipo Boolean, puesto que siempre será "Sí" o "No" (true si se cumple, y false si no se cumple).

## Ejercicio 4

Se pide desarrollar una función llamada "bust" que reciba una puntuación y que devuelva un Boolean:
- true si la puntuación es mayor que 21.
- false en caso contrario.

In [16]:
def bust(puntuacion: Int): Boolean = {
    puntuacion > 21
}

bust(18)
bust(21)
bust(22)
bust(30)

defined function bust
res16_1: Boolean = false
res16_2: Boolean = false
res16_3: Boolean = true
res16_4: Boolean = true

## Ejercicio 5

Se pide definir una función que devuelva el número mayor entre 2 números enteros

In [17]:
def maxHand(hand1: Int, hand2: Int) = {
    if (hand1 > hand2) hand1
    else hand2
}

maxHand(17, 19)
maxHand(20 , 18)
maxHand(21 , 21)

defined function maxHand
res17_1: Int = 19
res17_2: Int = 20
res17_3: Int = 21

### Explicación
Si ambos números tienen el mismo valor, se devuelve el valor de la mano 2, ya que al tener el mismo valor, es el mayor y a su vez el menor, por lo que está bien de todos modos.
En un futuro si se desea mejorar la función, se podría contemplar el devolver un valor nulo, lanzar una excepción, etc.